File: jeu/mesure.py
"""Mesure de reference : temps par phase + volume de frappe, par division."""
import random, statistics, sys
from collections import Counter
import engine
from engine import simuler_combat, DIVISIONS, ARMES_V2, TELEMETRY, reset_telemetry
from generator import generer_roster
from instrument import reset

def mesurer(n_par_div=60, seed=11, divisions=None, n_rosters=3):
    random.seed(seed)
    res = {}
    for div in (divisions or DIVISIONS):
        reset_telemetry()
        tent = touch = 0; n = 0
        issues = Counter()
        for _ in range(n_rosters):
            roster = generer_roster(40, division=div, niveau_min=45, niveau_max=85)
            for _ in range(n_par_div):
                a,b = random.sample(roster,2)
                fa,fb = a['fighter'], b['fighter']
                reset(fa); reset(fb)
                w, log = simuler_combat(fa,fb,rounds=3,verbose=False)
                t='\n'.join(log)
                if not w: issues["nul"]+=1
                elif 'tape !' in t: issues["SUB"]+=1
                elif 'TKO AU SOL' in t: issues["TKO sol"]+=1
                elif 'KO SEC' in t: issues["KO sec"]+=1
                elif 'TKO' in t: issues["TKO"]+=1
                else: issues["DEC"]+=1
                for l in log:
                    if "→ touché" in l: touch+=1; tent+=1
                    elif "→ manqué" in l: tent+=1
                    elif "ground and pound →" in l:
                        tent+=1
                        if "dégâts" in l: touch+=1
                n+=1
        minutes = TELEMETRY["t_total"]/60.0
        res[div] = {
            "n": n,
            "pc_debout": TELEMETRY["t_debout"]/TELEMETRY["t_total"]*100,
            "pc_clinch": TELEMETRY["t_clinch"]/TELEMETRY["t_total"]*100,
            "pc_sol":    TELEMETRY["t_sol"]/TELEMETRY["t_total"]*100,
            "min_par_combat": minutes/n,
            # /min = par minute de combat, mais pour UN combattant (donc /2)
            "tent_min": tent/minutes/2,
            "touch_min": touch/minutes/2,
            "prec": touch/max(1,tent)*100,
            "relances": TELEMETRY["n_relances"]/n,
            "td_tentes": TELEMETRY["td_tentes"]/n,
            "td_pc": TELEMETRY["td_reussis"]/max(1,TELEMETRY["td_tentes"])*100,
            "td_cl": TELEMETRY["td_clinch"]/n,
            "autres_seq": (TELEMETRY["sequences_sol"]-TELEMETRY["td_reussis"]-TELEMETRY["td_clinch"])/n,
            "seq_sol": TELEMETRY["sequences_sol"]/n,
            "duree_seq": TELEMETRY["t_sol"]/max(1,TELEMETRY["sequences_sol"]),
            "min_combat": TELEMETRY["t_total"]/60/n,
            "issues": issues,
            "ko": (issues["KO sec"]+issues["TKO"]+issues["TKO sol"])/max(1,n)*100,
        }
    return res

# ---------------------------------------------------------------------------
#  CIBLES REELLES — source : HonestElo / UFCStats, maj 1 aout 2026
#  Lues au MEME centile (30e des combattants actifs) dans chaque division,
#  correction indispensable : un top-25 sur 37 lourds actifs descend bien
#  plus bas dans le vivier qu'un top-25 sur 90 legers.
#
#  RESULTAT : le volume par minute est quasi PLAT (1,22x d'ecart) et NON
#  monotone avec le poids. Le lourd est le plus bas (8,1) mais la mouche
#  (8,8) est sous le leger (9,8). Le gradient reel est sur la PRECISION,
#  pas sur le volume : 49% en paille -> 58% en lourd, proprement monotone.
# ---------------------------------------------------------------------------
CIBLE_TENT = {  # frappes sig. tentees / minute de combat, par combattant
 "poids_paille":9.6,"poids_mouche":8.8,"poids_coq":9.1,"poids_plume":8.9,
 "poids_leger":9.8,"poids_welter":9.3,"poids_moyen":8.8,"poids_mi_lourd":9.4,
 "poids_lourd":8.1}

# Precision mesuree au 30e centile : 49 -> 58%. Le roster du moteur (45-85)
# couvre tout le vivier, pas seulement le tiers superieur : on garde la FORME
# du gradient en l'ancrant ~8 points plus bas, soit la moyenne UFC de 40-45%.
# Gradient de KO reel par division (8 591 combats UFC, deja valide - decision C8)
KO_REEL = {"poids_paille":13.5,"poids_mouche":24.1,"poids_coq":26.3,"poids_plume":29.7,
 "poids_leger":30.3,"poids_welter":33.7,"poids_moyen":38.4,"poids_mi_lourd":46.1,
 "poids_lourd":52.7}

CIBLE_PREC = {
 "poids_paille":41,"poids_mouche":42,"poids_coq":43,"poids_plume":44,
 "poids_leger":44,"poids_welter":45,"poids_moyen":45,"poids_mi_lourd":46,
 "poids_lourd":50}

if __name__ == "__main__":
    n = int(sys.argv[1]) if len(sys.argv)>1 else 60
    r = mesurer(n_par_div=n, n_rosters=int(sys.argv[2]) if len(sys.argv)>2 else 3)
    print(f"{'division':<15}{'debout':>7}{'clinch':>7}{'sol':>6}{'|':>2}"
          f"{'tent/m':>7}{'cib':>5}{'prec':>7}{'cib':>5}{'|':>2}{'TDtent':>7}{'seqSol':>7}{'dureeSeq':>9}{'relanc':>7}")
    print("-"*84)
    for d,v in r.items():
        print(f"{d:<15}{v['pc_debout']:>6.1f}%{v['pc_clinch']:>6.1f}%{v['pc_sol']:>5.1f}%{'|':>2}"
              f"{v['tent_min']:>7.1f}{CIBLE_TENT[d]:>5.1f}{v['prec']:>6.1f}%{CIBLE_PREC[d]:>5.0f}{'|':>2}"
              f"{v['td_tentes']:>7.1f}{v['td_pc']:>5.0f}%{v['td_cl']:>7.1f}{v['autres_seq']:>7.1f}{v['duree_seq']:>8.0f}s{v['ko']:>6.1f}{KO_REEL[d]:>6.1f}")
    print("\nCible : debout 65-70% | clinch ~10% | sol 20-25%")
    tot = Counter()
    for v in r.values(): tot.update(v["issues"])
    T = sum(tot.values())
    print("\nISSUES (reel : KO/TKO 32.6 | SUB 19.3 | DEC 46.8)")
    print("  " + " | ".join(f"{k} {v/T*100:.1f}%" for k,v in tot.most_common()))

File: jeu/js/mesure.js
/**
 * mesure.js — portage de mesure.py (le protocole de reference) et de
 * instrument.reset. C'est LUI qui prononce la bascule : il doit redonner
 *     DEC 47.6 | SUB 20.8 | TKO 19.4 | KO sec 10.9
 *
 * /!\ MISE A JOUR APRES LA BASCULE SCORING DU 08/08. L'ancienne ligne
 * affichait DEC 46.8 / nul 0.8 : elle etait devenue un FANTOME, on aurait
 * compare a un etat qui n'existait plus.
 * Le DELTA EST LA PREUVE que le calibrage est intact : SUB, TKO, KO sec et
 * TKO sol sont identiques AU DIXIEME (20.8 / 19.4 / 10.9 / 1.3). Seuls DEC
 * et nul ont bouge, de exactement 0.8 point, l'un vers l'autre — ce sont
 * les matchs nuls devenus des decisions, le 10-8 ayant quasi disparu. Le
 * scoring ne touche QUE les decisions, comme annonce.
 * avec mesurer(40, seed, 3) sur les graines 11 / 41 / 900.
 * L'ordre des tirages suit mesure.py : seed -> par division (ordre de
 * DIVISIONS) -> par roster (generer_roster) -> par combat (sample de 2).
 */

const { alea } = require("./alea.js");
const E = require("./engine.js");
const { generer_roster } = require("./generator.js");

/** instrument.reset : remise a neuf d'un combattant entre deux combats. */
function reset(x) {
  x.head_damage = 0; x.cardio = 100.0; x.sonne = 0; x.knockdowns = 0;
  x.legs.gauche = 0; x.legs.droite = 0;
  x.body.degats_corps = 0; x.body.degats_foie = 0;
  x.stance.garde_actuelle = x.stance.garde_naturelle;
}

function mesurer(n_par_div = 60, seed = 11, divisions = null, n_rosters = 3) {
  alea.seed(seed);
  const res = {};
  for (const div of (divisions || Object.keys(E.DIVISIONS))) {
    E.reset_telemetry();
    const issues = {};
    let n = 0;
    for (let r = 0; r < n_rosters; r++) {
      const roster = generer_roster(40, { division: div, niveau_min: 45, niveau_max: 85 });
      for (let c = 0; c < n_par_div; c++) {
        const [a, b] = alea.sample(roster, 2);
        const fa = a.fighter, fb = b.fighter;
        reset(fa); reset(fb);
        const [w, log] = E.simuler_combat(fa, fb, 3, false);
        const t = log.join("\n");
        let cle;
        if (!w) cle = "nul";
        else if (t.includes("tape !")) cle = "SUB";
        else if (t.includes("TKO AU SOL")) cle = "TKO sol";
        else if (t.includes("KO SEC")) cle = "KO sec";
        else if (t.includes("TKO")) cle = "TKO";
        else cle = "DEC";
        issues[cle] = (issues[cle] !== undefined ? issues[cle] : 0) + 1;
        n += 1;
      }
    }
    res[div] = { n, issues };
  }
  return res;
}

module.exports = { mesurer, reset };

// ------------------------------------------------------- execution directe
if (require.main === module) {
  const graines = process.argv.length > 2
    ? process.argv.slice(2).map(Number) : [11, 41, 900];
  const nParDiv = 40, nRosters = 3;
  const moyennes = {};
  for (const g of graines) {
    const r = mesurer(nParDiv, g, null, nRosters);
    const tot = {};
    let T = 0;
    for (const div of Object.keys(r))
      for (const [k, v] of Object.entries(r[div].issues)) {
        tot[k] = (tot[k] !== undefined ? tot[k] : 0) + v;
        T += v;
      }
    const ligne = ["DEC", "SUB", "TKO", "KO sec", "TKO sol", "nul"]
      .map(k => `${k} ${((tot[k] !== undefined ? tot[k] : 0) / T * 100).toFixed(1)}`);
    console.log(`  graine ${String(g).padStart(3)} (n=${T}) : ${ligne.join(" | ")}`);
    for (const k of ["DEC", "SUB", "TKO", "KO sec", "TKO sol", "nul"])
      (moyennes[k] = moyennes[k] || []).push((tot[k] !== undefined ? tot[k] : 0) / T * 100);
  }
  const moy = Object.entries(moyennes)
    .map(([k, v]) => `${k} ${(v.reduce((s, x) => s + x, 0) / v.length).toFixed(1)}`);
  console.log(`  MOYENNE ${graines.length} graines : ${moy.join(" | ")}`);
  console.log("  REFERENCE (chantier I complet, 09/08) : DEC 48.8 | SUB 23.8 | TKO 16.9 | KO sec 9.0 | TKO sol 1.4 | nul 0.0");
  console.log("  (08/08 scoring : DEC 47.6 | SUB 20.8 | TKO 19.4 | KO 10.9)");
  console.log("  /!\\ le calibrage a BOUGE, c'etait prevu : on repare les mecaniques,");
  console.log("      on recalibrera a la fin. Ne pas 'corriger' ce deplacement.");
}
